source/agora/crypto/Hash.d

```diff
@@ -194,6 +194,52 @@ public void hashPart (T) (scope const auto ref T record, scope HashDg hasher)
         static assert(0, "Unsupported type: " ~ T.stringof);
 }
 
+/*******************************************************************************
+
+    Call the provided delegate with a `HashDg`, allowing custom hash logic
+
+    This allow users to do some hash aggregation without relying on `hashMulti`,
+    but without the attributes restrictions of `hashPart`.
+
+    Params:
+      T = Type of predicate (delegate / function) to run
+      record = Instance of `T` to hash
+      state  = State delegate, when this struct is nested in another.
+
+    Returns:
+      The `Hash` representing this instance
+
+*******************************************************************************/
+
+public Hash hashPred (T) (scope T pred)
+{
+    static assert (is(T : void delegate (scope HashDg)) ||
+                   is(T : void function (scope HashDg)),
+                   T.stringof ~
+                   " does not implicitly convert to `void delegate (scope HashDg)`");
+
+    Hash hash = void;
+    crypto_generichash_state state;
+
+    void trustedInit () @trusted
+    {
+        crypto_generichash_init(&state, null, 0, Hash.sizeof);
+    }
+    trustedInit();
+
+    scope HashDg dg = (in ubyte[] data) @trusted {
+        crypto_generichash_update(&state, data.ptr, data.length);
+    };
+
+    pred(dg);
+    void trustedFini () @trusted
```

Comment:
Is `Fini` here short for "Final" ?

Reply:
Not quite, `.fini` is the section in the ELF binary where finalizers are stored, and I just reached for a short name. Could switch to something else, it's just really the first thing that came to mind.

Reply:
Read about the names. The names LGTM.